architecture/2024-05-radius-on-dapr.md

```diff
@@ -0,0 +1,324 @@
+
+I'm proposing that we require Dapr to by installed by users to develop Radius. We'll need the `daprd` binary to be present on the user's system.
+
+There are three tasks we need to solve to get Dapr working in development:
+
+- Configure statestore and pubsub components
+- Start the centralized "placement" service
+- Start a sidecar for each process
+
+I propose that we make UCP responsible for 1) and 2). These are tasks that should be done once for that are "system-wide". Inside UCP we'll write code that
+can generate the necessary component definitions on disk and will start the placement service. 
+
+For the components, we can start with SQLite for the state store and in-memory for pub-sub. This means that database changes will peristed across debugging sessions. Since SQLite is file-based it is easy to delete and recreate the database to recover from a bad state.
+
+Each service will also be responsible for the lifecycle of its own sidecar. 
+
+These behaviors will be configurable in our configuration file. This is just for development purposes and won't be used when Radius is running in Kubernetes.
+
+### API design (if applicable)
+
+*n/a: no API change.*
+
+### CLI Design (if applicable)
+
+*n/a: no CLI change.*
+
+### Implementation Details
+
+*n/a: described as part of detailed design.*
+
+### Error Handling
+
+No new error handling is needed in our existing. The data layer contract will remain the same.
+
+There are two possible cases for misconfiguration of Radius production scenarios:
+
+- Dapr has not been installed. The Dapr sidecars will not be injected, and so Radius will not be able to connect to Dapr. We should check for this condition at startup and crash with a descriptive error.
+- Dapr has been installed but no components are configured. We should use the Dapr metadata API to verify the configuration at startup and crash with a descriptive error. 
+
+## Test plan
+
+Dapr will become a required dependency that is installed in all Radius configurations. We will get significant test coverage from our existing functional and long-haul tests.
+
+Additionally we have an existing "compliance" suite of tests for our data store package. The new implementation of the Dapr state store data layer can leverage these tests.
+
+Dapr is also easy to use in integration tests because it provides in-memory components. We can test more of Radius in integration tests by using this support.
+
+## Security
+
+Adopting Dapr as described in this document does not change our existing security posture. However Dapr comes with many security features we can leverage in the future, and will make improving our security posture easier.
+
+## Compatibility (optional)
+
+We don't currently support upgrades from one release to another. Adopting Dapr and mature database technology like Postgres will help us support upgrades in the future.
```

Comment:
> Adopting Dapr and mature database technology like Postgres will help us support upgrades in the future

Curious to hear more on how this will help with upgrades

Reply:
I'd like to dig into this more in the detailed design doc for state storage.